File: rock63/src/main/java/com/uksusoff/rock63/ui/ItemListActivity.java
package com.uksusoff.rock63.ui;

import android.view.View;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;

import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.uksusoff.rock63.R;
import com.uksusoff.rock63.data.DataSource;
import com.uksusoff.rock63.exceptions.NoContentException;
import com.uksusoff.rock63.exceptions.NoInternetException;

import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;

@EActivity(R.layout.a_list)
public abstract class ItemListActivity extends BaseMenuActivity {

    @ViewById(R.id.list)
    ListView list;

    @ViewById(R.id.refresh)
    SwipeRefreshLayout refresh;

    @ViewById(R.id.list_empty)
    TextView emptyView;

    protected abstract boolean isRefreshing();

    protected abstract void setRefreshing(boolean refreshing);

    protected abstract ItemListActivity getActiveActivity();

    protected abstract void setActiveActivity(ItemListActivity activity);

    protected abstract ListAdapter createAdapterFromStorageItems();

    protected abstract void refreshItemStorage() throws NoInternetException, NoContentException;

    protected abstract int getEmptyListTextResId();

    @Override
    protected void init() {
        super.init();
        setActiveActivity(this);

        emptyView.setText(getEmptyListTextResId());

        loadNewsFromDatabase();

        refresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refreshList();
            }
        });
        refresh.post(new Runnable() {
            @Override
            public void run() {
                if (isRefreshing()) {
                    setRefreshIndicatorActive(true);
                } else if (list.getAdapter().isEmpty()) {
                    refreshList();
                }
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (getActiveActivity() == this) {
            setActiveActivity(null);
        }
    }

    public void refreshList() {
        setRefreshIndicatorActive(true);
        reloadAll();
    }

    @Background
    void reloadAll() {
        try {
            setRefreshing(true);
            refreshItemStorage();
            getActiveActivity().loadNewsFromDatabase();
        } catch (NoInternetException | NoContentException e) {
            showWarning(R.string.error_no_internet);
        } finally {
            setRefreshing(false);
        }
        setRefreshIndicatorActive(false);
    }

    @UiThread
    void setRefreshIndicatorActive(boolean active) {
        refresh.setRefreshing(active);
    }

    @UiThread
    void loadNewsFromDatabase() {

        list.setAdapter(createAdapterFromStorageItems());

        if (list.getAdapter().isEmpty()) {
            emptyView.setVisibility(View.VISIBLE);
        } else {
            emptyView.setVisibility(View.GONE);
        }
    }

}

File: rock63/src/main/java/com/uksusoff/rock63/ui/ItemListActivity.kt
package com.uksusoff.rock63.ui

import android.view.View
import android.widget.ArrayAdapter
import android.widget.ListAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.uksusoff.rock63.R
import com.uksusoff.rock63.exceptions.NoInternetException
import com.uksusoff.rock63.ui.adapters.AdvSimpleAdapter
import org.androidannotations.annotations.Background
import org.androidannotations.annotations.EActivity
import org.androidannotations.annotations.UiThread
import org.androidannotations.annotations.ViewById

/**
 * Created by Vyacheslav Vodyanov on 16.05.2016.
 */
@EActivity(R.layout.a_list)
abstract class ItemListActivity : BaseMenuActivity() {

    @ViewById(R.id.list)
    protected lateinit var list: ListView
    @ViewById(R.id.refresh)
    protected lateinit var refresh: SwipeRefreshLayout
    @ViewById(R.id.list_empty)
    protected lateinit var emptyView: TextView

    protected abstract fun createAdapterFromStorageItems(handler: (ListAdapter) -> Unit)
    protected abstract fun refreshItemStorage()
    protected abstract val emptyListTextResId: Int

    private var isTriedToLoad = false

    override fun init() {
        super.init()

        emptyView.setText(emptyListTextResId)
        loadItemsFromDatabase()

        refresh.setOnRefreshListener { reloadAll() }
    }

    @Background
    open fun reloadAll() {
        try {
            setRefreshIndicatorActive(true)
            refreshItemStorage()
        } catch (e: NoInternetException) {
            showWarning(R.string.error_no_internet)
        } finally {
            setRefreshIndicatorActive(false)
        }

        loadItemsFromDatabase()
    }

    @UiThread
    open fun setRefreshIndicatorActive(active: Boolean) {
        refresh.isRefreshing = active
    }

    @UiThread
    open fun loadItemsFromDatabaseInternal(adapter: ListAdapter) {
        list.adapter = adapter
        if (list.adapter.isEmpty) {
            if (!isTriedToLoad) {
                isTriedToLoad = true
                reloadAll()
            }
            emptyView.visibility = View.VISIBLE
        } else {
            emptyView.visibility = View.GONE
        }
    }

    @UiThread
    open fun loadItemsFromDatabase() {
        createAdapterFromStorageItems { loadItemsFromDatabaseInternal(it) }
    }
}